File: ChatGPT/Simples/dfs/Dfs.java
// Input  Using the previous graph, how to implement the Depth-First Search algorith?
import java.util.ArrayList;
import java.util.LinkedList;

public class Dfs {
    private ArrayList<LinkedList<Integer>> adjList;
    private int numVertices;

    public Dfs(int numVertices) {
        this.numVertices = numVertices;
        adjList = new ArrayList<>(numVertices);
        for (int i = 0; i < numVertices; i++) {
            adjList.add(new LinkedList<>());
        }
    }

    public void addEdge(int src, int dest) {
        adjList.get(src).add(dest);
        adjList.get(dest).add(src); // If the graph is undirected
    }

    public void dfs(int startVertex) {
        boolean[] visited = new boolean[numVertices];
        dfsRecursive(startVertex, visited);
    }

    private void dfsRecursive(int vertex, boolean[] visited) {
        visited[vertex] = true;
        System.out.print(vertex + " "); // Process the vertex

        for (int neighbor : adjList.get(vertex)) {
            if (!visited[neighbor]) { // Check if not visited
                dfsRecursive(neighbor, visited);
            }
        }
    }

    public static void main(String[] args) {
        Dfs g = new Dfs(4);
        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 3);
        g.dfs(0);
    }
}
/*
Score: 1
*/

File: ChatGPT/Simples/dfs/dfs.py
#Input: Using the previous graph, how to implement the Depth-First Search algorith?

class Graph:
    def __init__(self):
        self.graph = {}

    def add_vertex(self, vertex):
        if vertex not in self.graph:
            self.graph[vertex] = []

    def add_edge(self, u, v):
        if u not in self.graph:
            self.add_vertex(u)
        if v not in self.graph:
            self.add_vertex(v)
        self.graph[u].append(v)
        self.graph[v].append(u)  # For undirected graph

    def dfs(self, start_vertex, visited=None):
        if visited is None:
            visited = set()
        visited.add(start_vertex)
        print(start_vertex, end=' ')
        for neighbor in self.graph[start_vertex]:
            if neighbor not in visited:
                self.dfs(neighbor, visited)

    def __str__(self):
        return str(self.graph)

# Example usage
g = Graph()
g.add_vertex('A')
g.add_vertex('B')
g.add_vertex('C')
g.add_vertex('D')
g.add_edge('A', 'B')
g.add_edge('A', 'D')
g.add_edge('B', 'C')

print("DFS traversal starting from vertex 'A':")
g.dfs('A')

"""
Output:
    A B C D

Score: 1
"""